src/Cardano/Shell/NodeIPC/Lib.hs

```diff
@@ -139,61 +161,112 @@ getIPCHandle = do
             Left err -> throwM $ UnableToParseNodeChannel $ toS err
             Right fd -> liftIO $ fdToHandle fd
 
--- | Start NodeJS IPC with given 'ReadHandle', 'WriteHandle' and 'Port'
-startNodeJsIPC :: (MonadIO m) => ReadHandle -> WriteHandle -> Port -> m ()
-startNodeJsIPC readHandle writeHandle port =
+-- | Start IPC with given 'ReadHandle', 'WriteHandle' and 'Port'
+startIPC :: forall m. (MonadIO m) => ReadHandle -> WriteHandle -> Port -> m ()
+startIPC readHandle writeHandle port = liftIO $ void $ ipcListener readHandle writeHandle port
+
+-- | Start IPC with NodeJS
+--
+-- This only works if NodeJS spawns the Haskell executable as child process
+-- (See @server.js@ as an example)
+startNodeJsIPC :: forall m. (MonadIO m) => Port -> m ()
+startNodeJsIPC port = do
+    handle <- liftIO $ getIPCHandle
+    let readHandle = ReadHandle handle
+    let writeHandle = WriteHandle handle
     liftIO $ void $ ipcListener readHandle writeHandle port
 
 -- | Start IPC listener with given Handles and Port
 --
--- when the listener recieves 'Ping' it will return 'Pong'.
+-- When the listener recieves 'Ping' it will return 'Pong'.
 --
 -- If it recieves 'QueryPort', then the listener
 -- responds with 'ReplyPort' with 'Port',
-ipcListener :: forall m . (MonadIO m, MonadCatch m) => ReadHandle -> WriteHandle -> Port -> m ()
-ipcListener readHndl@(ReadHandle rHndl) writeHndl@(WriteHandle wHndl) (Port port) = do
-    catches handleMsgIn [Handler handler, Handler handleMsgError]
+ipcListener :: forall m . (MonadIO m, MonadCatch m, MonadMask m) => ReadHandle -> WriteHandle -> Port -> m ()
+ipcListener readHandle@(ReadHandle rHndl) writeHandle@(WriteHandle wHndl) (Port port) =
+    finally
+        (do
+            checkHandles readHandle writeHandle
+            catches handleMsgIn [Handler handler, Handler handleMsgError]
+        )
+        shutdown
   where
     handleMsgIn :: m ()
     handleMsgIn = do
         liftIO $ hSetNewlineMode rHndl noNewlineTranslation
         send Started
-        msgIn <- readMessage readHndl
+        msgIn <- readMessage readHandle
         case msgIn of
-            QueryPort -> send (ReplyPort port) >> shutdown
-            Ping      -> do
-                send Pong
-                shutdown
+            QueryPort          -> send (ReplyPort port)
+            Ping               -> send Pong
             -- TODO:Handle them nicely
-            MessageInFailure _ -> shutdown
+            MessageInFailure _ -> return ()
```

Comment:
If the message that gets in fails, won't the sender of that message wait for it's response?
And if it is waiting for the response, won't that simply leave the initial process hanging?
Wouldn't it be nicer if the _sender_ got some response to know that things went bad?
`MessageOutFailure` with the wrong format or something similar?

Reply:
I'm not understanding this comment well. This is when the sender is sending `MessageInFailure` message to the IPC so isn't it up to the sender what to do with their own process when they send this message to the node?

Reply:
Whops. Yes, you are correct, this is the message from the server.
But the message is saying "the last message I got was invalid"?
I guess this is good for now, since we really don't know what to do with that information.